### PR TITLE
[Model]Support MiniCPM-V 4.5

### PR DESCRIPTION
**Pull request by tc-mb**
## Description

Add MiniCPM-V 4.5 multimodal model support to the PyTorch backend, enabling
text + image + video inference via trtllm-serve.

The implementation follows the existing multimodal model pattern (Llava, Qwen3-VL, etc.):

- **Vision encoder**: SigLIP-based NaViT model with multi-resolution support
- **Resampler**: Cross-attention based vision-to-LLM feature projection
- **LLM**: Reuses the Qwen3-style language model backbone (checkpoint prefix normalization in the weight mapper)
- **Input processor**: Handles mixed text/image/video prompts, video temporal embeddings, and placeholder expansion
- **Video support**: Frame sampling with configurable max frames (default 180), temporal positional encoding, and multi-frame packing (default 3)

### Changes

| File | Description |
|------|-------------|
| `tensorrt_llm/_torch/models/modeling_minicpmv.py` | Main VLM implementation (vision encoder, resampler, conditional generation model) |
| `tensorrt_llm/_torch/models/checkpoints/hf/minicpmv_weight_mapper.py` | HF checkpoint weight mapping with `llm.` prefix normalization for the Qwen3 backbone |
| `tensorrt_llm/_torch/models/__init__.py` | Register `MiniCPMVForConditionalGeneration` |
| `tensorrt_llm/inputs/multimodal.py` | Propagate `temporal_ids` for video input token length calculation |

## PR Checklist

- [ ] Please check this after reviewing the above items as appropriate for this PR.

## Summary by CodeRabbit

* **New Features**
  * Added support for MiniCPM-V models, including multimodal image/video input handling and generation.
  * Expanded available model exports so MiniCPM-V can be imported directly.
  * Added checkpoint loading support for MiniCPM-V weights.

* **Bug Fixes**
  * Improved video token-length calculation by passing along temporal information when present, helping multimodal prompts map more accurately.

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_minicpmv.py [hunk 1]
```diff
@@ -0,0 +1,1230 @@
...
+        dtype = vision_hidden_states.dtype
+        patch_len = tgt_sizes[:, 0] * tgt_sizes[:, 1]
+        self._adjust_pos_cache(tgt_sizes, device=device)
+
+        temporal_ids_flatten = None
+        temporal_pos_emb = False
+        if temporal_ids is not None:
+            temporal_ids_flatten = list(chain.from_iterable(temporal_ids))
+            max_temporal_size = max(temporal_ids_flatten) + 1
+            if max_temporal_size > -1:
```
Comment (aswinvisva) [inline tensorrt_llm/_torch/models/modeling_minicpmv.py:866]: This check seems to not guard anything iiuc, `temporal_ids` max is -1, then max_temporal_size always becomes at least 0 , and 0 > -1, so this bool `temporal_pos_emb` is always True, no?

File: tensorrt_llm/inputs/multimodal.py [hunk 2]
```diff
@@ -1021,6 +1035,10 @@ def find_mm_token_lengths(
                     call_kwargs["video_grid_thw"] = (
                         video_grid_thw_for_items if len(items) == 1 else
                         video_grid_thw_for_items[idx:idx + 1])
+                if video_temporal_ids_for_items is not None:
```
Comment (aswinvisva) [inline tensorrt_llm/inputs/multimodal.py:1038]: please check that this works for multiple videos, iiuc the `get_num_tokens_per_video` expects a `list[list[int]]` while this returns a flattened `list[int]`

File: tensorrt_llm/_torch/models/modeling_minicpmv.py [hunk 3]
```diff
@@ -0,0 +1,1230 @@
...
+        llm_model_config.extra_attrs = model_config.extra_attrs
+        self.llm = AutoModelForCausalLM.from_config(llm_model_config)
+
+        self.mm_encoder = None
+        if not _is_mm_disagg():
+            self.mm_encoder = MiniCPMVVisionModel(copy.deepcopy(model_config)).eval()
+
+        self.post_config()
+
+    def post_config(self) -> None:
```
Comment (aswinvisva) [inline tensorrt_llm/_torch/models/modeling_minicpmv.py:1126]: this will replace original MiniCPMV config with the Qwen3 LLM config silently, check that this doesn't break anything or cause any issues downstream in the future when trying to access properties from the MiniCMPV config

File: tensorrt_llm/_torch/models/modeling_minicpmv.py [hunk 4]
```diff
@@ -0,0 +1,1230 @@
...
+    def load_weights(self, weights: dict[str, torch.Tensor]) -> None:
+        self.load_state_dict(_filter_weights(weights, "resampler."),
+                             strict=True)
+        self.to(dtype=self.dtype)
+
+
+class MiniCPMVVisionModel(nn.Module):
+    """MiniCPM-V vision encoder plus resampler."""
+
+    def __init__(self, model_config: ModelConfig[PretrainedConfig]) -> None:
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_minicpmv.py:919]: This signature breaks encoder-only mode. When `mm_encoder_only=True` (EPD disaggregated serving), `AutoModelForCausalLM.from_config` constructs the registered vision encoder as `vision_encoder_cls(config, vlm_base_model)` — two positional args (see `modeling_auto.py:59`). `MiniCPMVVisionModel.__init__` only accepts `model_config`, so encoder-only startup raises `TypeError` for this model. Add `*args, **kwargs` (as `LlavaNextVisionModel` does at `modeling_llava_next.py:435`) or accept the base-model argument explicitly.

File: tensorrt_llm/_torch/models/modeling_minicpmv.py [hunk 5]
```diff
@@ -0,0 +1,1230 @@
...
+        self,
+        video: Any,
+        frames: list[Any],
+        mm_processor_kwargs: dict[str, Any],
+    ) -> Optional[list[list[int]]]:
+        enable_temporal_ids = mm_processor_kwargs.pop(
+            "enable_video_temporal_ids",
+            mm_processor_kwargs.pop("use_video_temporal_ids", False),
+        )
+        if not enable_temporal_ids:
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_minicpmv.py:313]: When `enable_video_temporal_ids` is unset (the default) this early return skips the `pop`s below, so `video_max_num_frames` / `video_max_num_packing` / `video_force_packing` / `video_time_scale` stay in `mm_processor_kwargs` and get splatted into the HF processor call in `_preprocess` (line 390), which can fail on unexpected kwargs. Same when the caller passes explicit `temporal_ids` (this function is never called). Pop all the `video_*` knobs unconditionally before this check.

File: tensorrt_llm/_torch/models/modeling_minicpmv.py [hunk 6]
```diff
@@ -0,0 +1,1230 @@
...
+                                 self.embed_dim,
+                                 bias=False)
+        self.attn = nn.MultiheadAttention(self.embed_dim, self.num_heads)
+        self.ln_q = nn.LayerNorm(self.embed_dim, eps=1e-6)
+        self.ln_kv = nn.LayerNorm(self.embed_dim, eps=1e-6)
+        self.ln_post = nn.LayerNorm(self.embed_dim, eps=1e-6)
+        self.proj = nn.Parameter((self.embed_dim**-0.5) *
+                                 torch.randn(self.embed_dim, self.embed_dim))
+        self._set_2d_pos_cache(self.max_size)
+        self._set_temporal_pos_cache(self.max_temporal_size)
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_minicpmv.py:741]: This builds a 72000 × hidden_size float32 table at init (~1.2 GB host, then bf16 after the `.to(dtype)` below) and registers it as a module buffer, so it moves to GPU with the model — ~590 MB of VRAM at hidden_size 4096, paid even for image-only serving. Consider keeping `temporal_pos_embed` on CPU and gathering only the rows for the temporal ids used in each forward, or building the cache lazily up to the max temporal id actually observed.

File: tensorrt_llm/_torch/models/modeling_minicpmv.py [hunk 7]
```diff
@@ -0,0 +1,1230 @@
...
+                                    _DEFAULT_VIDEO_TIME_SCALE))
+
+        if max_num_frames <= 0:
+            raise ValueError("video_max_num_frames must be positive.")
+        if max_num_packing <= 0:
+            raise ValueError("video_max_num_packing must be positive.")
+        if time_scale <= 0:
+            raise ValueError("video_time_scale must be positive.")
+
+        packing_nums = min(
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_minicpmv.py:334]: `video_max_num_frames` never limits the frame count — all frames are still sent to the processor; it only raises `packing_nums` so more frames share one temporal group. The PR description advertises "frame sampling with configurable max frames (default 180)", which this doesn't do. If a >180-frame video should be subsampled (as in the reference implementation), that logic is missing; otherwise rename/document the knob so users don't expect it to cap cost.

File: tensorrt_llm/_torch/models/modeling_minicpmv.py [hunk 8]
```diff
@@ -0,0 +1,1230 @@
...
+
+    def _get_requests_with_mm_data(
+        self,
+        multimodal_params: list[MultimodalParams],
+    ) -> list[MultimodalParams]:
+        mm_multimodal_params = []
+        for multimodal_param in multimodal_params:
+            data = multimodal_param.multimodal_data
+            if (
+                data.get("image", {}).get("pixel_values") is not None
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_minicpmv.py:1200]: `data.get("image", {}).get("pixel_values")` raises `AttributeError` if `multimodal_data` contains `"image": None` (key present, value None). `_parse_and_batch_multimodal_data` guards this case with `data.get("image") or data.get("video") or {}` — use the same pattern here for consistency.
